### DISCUSSION TITLE
webpack providePlugin

### DISCUSSION THREAD

**Discussion by wangcongyi** [webpack/webpack#15631] [Q&A]
Accodrding to the webpack doc.   
providePlugin can `Automatically load modules instead of having to import or require them everywhere`   
so can i `put` all the modules in providePlugin ? ( include `react` `redux` `react-redux`. )
what situation or condition can put modules in providePlugin ?
what difference bettween put in providePlugin and not after build ?

**Comment by vankop** [comment]
`providePlugin` is good for kind of polyfilling.. in general it is the same as `import x from "x"` difference is that you don't need to change code ( adding imports ). Simple example:
index.js
```js
const url = new URL(...);
```
webpack.config.js
```js
module.exports = {
  plugins: [
     new ProvidedPlugin({ URL: "./my-url" })
  ]
};
```

so resulted code would looks like
```js
import URL from "./my-url";
const url = new URL(...);
```

**Reply by vankop** [reply]
> can i put all the modules in providePlugin ? ( include react redux react-redux. )

You can, but there is no tree-shaking for this.

**Comment by wangcongyi** [comment]
@vankop  Maybe i need more information. 
 Can i think  at the `providePlugin ` modules, webpack put the module at `top level closure` after build ?
How does work  ` no tree-shaking for this` ?
tks your reply

**Reply by vankop** [reply]
right now ProvidedPlugin does not have tree-shaking e.g. you provide module with 3 exports but use only one. right now all exports will be bundled
